File: dotNET/PFProSDK/src/PayPal/Payments/DataObjects/PayflowConnectionData.cs
#region "Imports"

using System;
using PayPal.Payments.Common.Utility;
using PayPal.Payments.Common;

#endregion

namespace PayPal.Payments.DataObjects
{
	/// <summary>
	/// Used for Payflow Host related information.
	/// </summary>
	/// <remarks>This class stores the information related to connection to the 
	/// PayPal payment gateway. If the empty constructor of this class 
	/// is used to create the object, or 
	/// passed values are empty, then The following values (if empty) are looked for 
	/// as follows:
	/// <list type="table">
	/// <listheader>
	/// <term>Property</term>
	/// <description>From Internal Default</description>
	/// <description>From App.config key</description>
	/// </listheader>
	/// <item>
	/// <term>Payflow Host</term>
	/// <description>NA</description>
	/// <description>PAYFLOW_HOST</description>
	/// </item>
	/// <item>
	/// <term>Payflow Port</term>
	/// <description>443</description>
	/// <description>NA</description>
	/// </item>
	/// <item>
	/// <term>Transaction timeout</term>
	/// <description>45 seconds</description>
	/// <description>NA</description>
	/// </item>
	/// </list>
	/// </remarks>
	public sealed class PayflowConnectionData : BaseRequestDataObject
	{
		#region "Member Variables"

		/// <summary>
		/// Payflow Host address
		/// </summary>
		private String mHostAddress;

		/// <summary>
		/// Payflow Host port
		/// </summary>
		private int mHostPort;

		/// <summary>
		/// Proxy Address
		/// </summary>
		private String mProxyAddress;

		/// <summary>
		/// Proxy Port
		/// </summary>
		private int mProxyPort;

		/// <summary>
		/// Proxy Logon Id
		/// </summary>
		private String mProxyLogon;

		/// <summary>
		/// Proxy Password
		/// </summary>
		private String mProxyPassword;

		/// <summary>
		/// Transaction TimeOut
		/// </summary>
		private int mTimeOut;

		#endregion

		#region "Properties"

		/// <summary>
		/// Gets HostAddress. It is PayPal's HostName
		/// </summary>
		/// <remarks>Read-only property.</remarks>
		public String HostAddress
		{
			get { return mHostAddress; }
		}

		/// <summary>
		/// Gets HostPort. Use port 443
		/// </summary>
		/// <remarks>Read-only property.</remarks>
		public int HostPort
		{
			get { return mHostPort; }
		}

		/// <summary>
		/// Gets Time-out period for the transaction. The minimum recommended
		/// time-out value is 30 seconds. The client begins tracking
		/// from the time that it sends the transaction request to the server.
		/// </summary>
		/// <remarks>Read-only property.</remarks>
		public int TimeOut
		{
			get { return mTimeOut; }
		}

		/// <summary>
		/// Gets Proxy server address. Use the PROXY parameters for servers
		/// behind a firewall. Your network administrator can provide the
		/// values.
		/// </summary>
		/// <remarks>Read-only property.</remarks>
		public String ProxyAddress
		{
			get { return mProxyAddress; }
		}

		/// <summary>
		/// Gets ProxyPort
		/// </summary>
		/// <remarks>Read-only property.</remarks>
		public int ProxyPort
		{
			get { return mProxyPort; }
		}

		/// <summary>
		/// Gets ProxyLogon
		/// </summary>
		/// <remarks>Read-only property.</remarks>
		public String ProxyLogon
		{
			get { return mProxyLogon; }
		}

		/// <summary>
		/// Gets ProxyPassword
		/// </summary>
		/// <remarks>Read-only property.</remarks>
		public String ProxyPassword
		{
			get { return mProxyPassword; }
		}

		#endregion

		#region "Constructors"

		/// <summary>
		/// Constructor
		/// </summary>
		/// <remarks>
		/// The following values (if empty) are looked for 
		/// as follows:
		/// <list type="table">
		/// <listheader>
		/// <term>Property</term>
		/// <description>From Internal Default</description>
		/// <description>From App.config key</description>
		/// </listheader>
		/// <item>
		/// <term>Payflow Host</term>
		/// <description>NA</description>
		/// <description>Payflow_HOST</description>
		/// </item>
		/// <item>
		/// <term>Payflow Port</term>
		/// <description>443</description>
		/// <description>NA</description>
		/// </item>
		/// <item>
		/// <term>Transaction timeout</term>
		/// <description>45 seconds</description>
		/// <description>NA</description>
		/// </item>
		/// </list>
		/// </remarks>
		public PayflowConnectionData() : this(null, 0, 0, null, 0, null, null)
		{
		}

		/// <summary>
		/// Constructor
		/// </summary>
		/// <param name="HostAddress">Payflow Host Address</param>
		/// <remarks>
		/// The following values (if empty) are looked for 
		/// as follows:
		/// <list type="table">
		/// <listheader>
		/// <term>Property</term>
		/// <description>From Internal Default</description>
		/// <description>From App.config key</description>
		/// </listheader>
		/// <item>
		/// <term>Payflow Host</term>
		/// <description>NA</description>
		/// <description>Payflow_HOST</description>
		/// </item>
		/// <item>
		/// <term>Payflow Port</term>
		/// <description>443</description>
		/// <description>NA</description>
		/// </item>
		/// <item>
		/// <term>Transaction timeout</term>
		/// <description>45 seconds</description>
		/// <description>NA</description>
		/// </item>
		/// </list>
		/// </remarks>
		public PayflowConnectionData(String HostAddress) : this(HostAddress, 0, 0, null, 0, null, null)
		{
		}

		/// <summary>
		/// Constructor
		/// </summary>
		/// <param name="HostAddress">Payflow Host Address</param>
		/// <param name="HostPort">Payflow Host port</param>
		/// <param name="TimeOut">Transaction time out</param>
		/// <remarks>
		/// The following values (if empty) are looked for 
		/// as follows:
		/// <list type="table">
		/// <listheader>
		/// <term>Property</term>
		/// <description>From Internal Default</description>
		/// <description>From App.config key</description>
		/// </listheader>
		/// <item>
		/// <term>Payflow Host</term>
		/// <description>NA</description>
		/// <description>Payflow_HOST</description>
		/// </item>
		/// <item>
		/// <term>Payflow Port</term>
		/// <description>443</description>
		/// <description>NA</description>
		/// </item>
		/// <item>
		/// <term>Transaction timeout</term>
		/// <description>45 seconds</description>
		/// <description>NA</description>
		/// </item>
		/// </list>
		/// </remarks>		
		public PayflowConnectionData(String HostAddress, int HostPort, int TimeOut) : this(HostAddress, HostPort, TimeOut, null, 0, null, null)
		{
		}

		/// <summary>
		/// Constructor
		/// </summary>
		/// <param name="HostAddress">Payflow Host Address</param>
		/// <param name="HostPort">>Payflow Host port</param>
		/// <param name="TimeOut">Transaction timeout</param>
		/// <param name="ProxyAddress">Proxy Address</param>
		/// <param name="ProxyPort">Proxy Port</param>
		/// <param name="ProxyLogon">Proxy Logon Id </param>
		/// <param name="ProxyPassword">ProxyPwd</param>
		/// <remarks>
		/// The following values (if empty) are looked for 
		/// as follows:
		/// <list type="table">
		/// <listheader>
		/// <term>Property</term>
		/// <description>From Internal Default</description>
		/// <description>From App.config key</description>
		/// </listheader>
		/// <item>
		/// <term>Payflow Host</term>
		/// <description>NA</description>
		/// <description>Payflow_HOST</description>
		/// </item>
		/// <item>
		/// <term>Payflow Port</term>
		/// <description>443</description>
		/// <description>NA</description>
		/// </item>
		/// <item>
		/// <term>Transaction timeout</term>
		/// <description>45 seconds</description>
		/// <description>NA</description>
		/// </item>
		/// </list>
		/// </remarks>
		public PayflowConnectionData(String HostAddress, int HostPort, int TimeOut, String ProxyAddress, int ProxyPort, String ProxyLogon, String ProxyPassword)
		{
			if (Context == null)
			{
				Context = new Context();
			}

			mHostAddress = HostAddress;
			mHostPort = HostPort;
			mTimeOut = TimeOut;
			mProxyAddress = ProxyAddress;
			mProxyPort = ProxyPort;
			mProxyLogon = ProxyLogon;
			mProxyPassword = ProxyPassword;
			InitDefaultValues();
		}

		/// <summary>
		/// Constructor
		/// </summary>
		/// <param name="HostAddress">Payflow Host Address</param>
		/// <param name="HostPort">>Payflow Host port</param>
		/// <remarks>
		/// The following values (if empty) are looked for 
		/// as follows:
		/// <list type="table">
		/// <listheader>
		/// <term>Property</term>
		/// <description>From Internal Default</description>
		/// <description>From App.config key</description>
		/// </listheader>
		/// <item>
		/// <term>Payflow Host</term>
		/// <description>NA</description>
		/// <description>Payflow_HOST</description>
		/// </item>
		/// <item>
		/// <term>Payflow Port</term>
		/// <description>443</description>
		/// <description>NA</description>
		/// </item>
		/// <item>
		/// <term>Transaction timeout</term>
		/// <description>45 seconds</description>
		/// <description>NA</description>
		/// </item>
		/// </list>
		/// </remarks>		
		public PayflowConnectionData(String HostAddress, int HostPort) : this(HostAddress, HostPort, 0, null, 0, null, null)
		{
		}

		/// <summary>
		/// Constructor
		/// </summary>
		/// <param name="HostAddress">Payflow Host Address</param>
		/// <param name="HostPort">>Payflow Host port</param>
		/// <param name="ProxyAddress">Proxy Address</param>
		/// <param name="ProxyPort">Proxy Port</param>
		/// <param name="ProxyLogon">Proxy Logon Id </param>
		/// <param name="ProxyPassword">ProxyPwd</param>
		/// <remarks>
		/// The following values (if empty) are looked for 
		/// as follows:
		/// <list type="table">
		/// <listheader>
		/// <term>Property</term>
		/// <description>From Internal Default</description>
		/// <description>From App.config key</description>
		/// </listheader>
		/// <item>
		/// <term>Payflow Host</term>
		/// <description>NA</description>
		/// <description>Payflow_HOST</description>
		/// </item>
		/// <item>
		/// <term>Payflow Port</term>
		/// <description>443</description>
		/// <description>NA</description>
		/// </item>
		/// <item>
		/// <term>Transaction timeout</term>
		/// <description>45 seconds</description>
		/// <description>NA</description>
		/// </item>
		/// </list>
		/// </remarks>		
		public PayflowConnectionData(String HostAddress, int HostPort, String ProxyAddress, int ProxyPort, String ProxyLogon, String ProxyPassword) : this(HostAddress, HostPort, 0, ProxyAddress, ProxyPort, ProxyLogon, ProxyPassword)
		{
		}

		#endregion

		/// <summary>
		/// Initializes the default values
		/// </summary>
		private void InitDefaultValues()
		{
			//Check if the values held
			//in the PayPal server
			//connection related params
			//if they are passed null or
			//0 (for int values) then
			//initialize them to appropriate
			//default values.

			//set the timeout to default timeout.
            if (mTimeOut == 0)
            {
                mTimeOut = PayflowConstants.DEFAULT_TIMEOUT;
            }
            
            try
            {
				if (mHostAddress == null || mHostAddress.Length == 0)
				{
					String HostAddress = PayflowUtility.AppSettings(PayflowConstants.INTL_PARAM_PAYFLOW_HOST);
					if (HostAddress != null && HostAddress.Length > 0)
					{
						HostAddress = HostAddress.TrimStart().TrimEnd();
						if (HostAddress.Length == 0)
						{
							String RespMessage = PayflowConstants.PARAM_RESULT
								+ PayflowConstants.SEPARATOR_NVP
								+ (String)PayflowConstants.CommErrorCodes[PayflowConstants.E_CONFIG_ERROR]
								+ PayflowConstants.DELIMITER_NVP
								+ PayflowConstants.PARAM_RESPMSG
								+ PayflowConstants.SEPARATOR_NVP
								+ (String)PayflowConstants.CommErrorMessages[PayflowConstants.E_CONFIG_ERROR]
								+ "Tag "
								+ PayflowConstants.INTL_PARAM_PAYFLOW_HOST +
								" is not present in the config file or config file is missing.";
							ErrorObject Error = new ErrorObject(PayflowConstants.SEVERITY_FATAL, "",RespMessage);
							Context.AddError(Error);
						}
						else
						{
							mHostAddress = HostAddress;
						}
					}
					else
					{
						String RespMessage = PayflowConstants.PARAM_RESULT
							+ PayflowConstants.SEPARATOR_NVP
							+ (String)PayflowConstants.CommErrorCodes[PayflowConstants.E_CONFIG_ERROR]
							+ PayflowConstants.DELIMITER_NVP
							+ PayflowConstants.PARAM_RESPMSG
							+ PayflowConstants.SEPARATOR_NVP
							+ (String)PayflowConstants.CommErrorMessages[PayflowConstants.E_CONFIG_ERROR]
							+ "Tag "
							+ PayflowConstants.INTL_PARAM_PAYFLOW_HOST +
							" is not present in the config file or config file is missing.";
						ErrorObject Error = new ErrorObject(PayflowConstants.SEVERITY_FATAL, "",RespMessage);
						Context.AddError(Error);
					}
				}
			}
			catch (Exception Ex)
			{
				String StackTrace = PayflowConstants.EMPTY_STRING;
				PayflowUtility.InitStackTraceOn();
				if(PayflowConstants.TRACE_ON.Equals(PayflowConstants.TRACE))
				{
					StackTrace = ": " + Ex.Message + Ex.StackTrace;
				}
				String RespMessage = PayflowConstants.PARAM_RESULT
					+ PayflowConstants.SEPARATOR_NVP
					+ (String)PayflowConstants.CommErrorCodes[PayflowConstants.E_CONFIG_ERROR]
					+ PayflowConstants.DELIMITER_NVP
					+ PayflowConstants.PARAM_RESPMSG
					+ PayflowConstants.SEPARATOR_NVP
					+ (String)PayflowConstants.CommErrorMessages[PayflowConstants.E_CONFIG_ERROR]
					+ "Tag "
					+ PayflowConstants.INTL_PARAM_PAYFLOW_HOST +
					" is not present in the config file or config file is missing."
					+ StackTrace;
				ErrorObject Error = new ErrorObject(PayflowConstants.SEVERITY_FATAL, "",RespMessage);
				Context.AddError(Error);
			}
		}
	}
}

File: java/src/sdk/dataobjects/paypal/payflow/PayflowConnectionData.java
package paypal.payflow;

/**
 * Used for Payflow Host related information.
 * <P>This class stores the information related to connection to the
 * PayPal payment gateway. If the empty constructor of this class
 * is used to create the object, or
 * passed values are empty, then The following values (if empty) are looked for
 * as follows:
 *
 * Property From Internal Default From SDK Properties
 * Payflow Host - NA - HostAddress
 * Payflow Port- 443 - HostPort
 * Transaction timeout - 45 seconds - NA
 *
 */
public final class PayflowConnectionData extends BaseRequestDataObject {

    private String hostAddress;
    private int hostPort;
    private String proxyAddress;
    private int proxyPort;
    private String proxyLogon;
    private String proxyPassword;
    private int timeOut;

    /**
     * Gets HostAddress. It is PayPal's HostName
     *
     * @return String
     */
    public String getHostAddress() {
        return hostAddress;
    }

    /**
     * Gets the Host port.Port 443 is used.
     *
     * @return int
     */
    public int getHostPort() {
        return hostPort;
    }

    /**
     * Gets Proxy server address. Use the PROXY parameters for servers
     * behind a firewall. Your network administrator can provide the
     * values.
     *
     * @return String
     */
    public String getProxyAddress() {
        return proxyAddress;
    }

    /**
     * Gets the proxy port.
     *
     * @return int
     */
    public int getProxyPort() {
        return proxyPort;
    }

    /**
     * Gets proxy Logon.
     *
     * @return String
     */
    public String getProxyLogon() {
        return proxyLogon;
    }

    /**
     * Gets the proxy password.
     *
     * @return String
     */
    public String getProxyPassword() {
        return proxyPassword;
    }

    /**
     * Gets Time-out period for the transaction. The minimum recommended
     * time-out value is 30 seconds. The client begins tracking
     * from the time that it sends the transaction request to the server.
     *
     * @return int
     */
    public int getTimeOut() {
        return timeOut;
    }

    /**
     * Constructor
     */
    public PayflowConnectionData() {
        this(null, 0, 0, null, 0, null, null);
    }

    /**
     * Constructor
     *
     * @param HostAddress String
     */
    public PayflowConnectionData(String HostAddress) {
        this(HostAddress, 0, 0, null, 0, null, null);
    }

    /**
     * Constructor
     *
     * @param HostAddress String
     * @param HostPort    Integer
     * @param Timeout     Integer
     */
    public PayflowConnectionData(String HostAddress, int HostPort, int Timeout) {
        this(HostAddress, HostPort, Timeout, null, 0, null, null);
    }

    /**
     * Constructor
     *
     * @param HostAddress   String
     * @param HostPort      Integer
     * @param TimeOut       Integer
     * @param ProxyAddress  String
     * @param ProxyPort     Integer
     * @param ProxyLogon    String
     * @param ProxyPassword String
     */
    public PayflowConnectionData(String HostAddress, int HostPort, int TimeOut, String ProxyAddress, int ProxyPort, String ProxyLogon, String ProxyPassword) {
        if (getContext() == null) {
            this.setContext(new Context());
        }

        hostAddress = HostAddress;
        hostPort = HostPort;
        timeOut = TimeOut;
        proxyAddress = ProxyAddress;
        proxyPort = ProxyPort;
        proxyLogon = ProxyLogon;
        proxyPassword = ProxyPassword;
        this.initSDKProperties();
    }

    /**
     * Constructor
     *
     * @param HostAddress String
     * @param HostPort    Integer
     */
    public PayflowConnectionData(String HostAddress, int HostPort) {
        this(HostAddress, HostPort, 0, null, 0, null, null);
    }

    /**
     * Constructor
     *
     * @param HostAddress   String
     * @param HostPort      Integer
     * @param ProxyAddress  String
     * @param ProxyPort     Integer
     * @param ProxyLogon    String
     * @param ProxyPassword String
     */
    public PayflowConnectionData(String HostAddress, int HostPort, String ProxyAddress, int ProxyPort, String ProxyLogon, String ProxyPassword) {
        this(HostAddress, HostPort, 0, ProxyAddress, ProxyPort, ProxyLogon, ProxyPassword);
    }

    /**
     * For any value not passed in the constructor the values being set in the properties will be used.
     */
    private void initSDKProperties() {
        Logger.getInstance().log("paypal.payflow.PayflowConnectionData.initSDKProperties(): Entered", PayflowConstants.SEVERITY_DEBUG);
        if (timeOut == 0) {
            timeOut = SDKProperties.getTimeOut();
        }
        if (hostPort == 0) {
            hostPort = SDKProperties.getHostPort();
        }
        if (null == hostAddress || hostAddress.trim().length() == 0) {
            hostAddress = SDKProperties.getHostAddress();
            if (null == hostAddress || hostAddress.trim().length() == 0) {
                String RespMessage = PayflowConstants.PARAM_RESULT
                        + PayflowConstants.SEPARATOR_NVP
                        + PayflowConstants.CommErrorCodes.get(PayflowConstants.E_INIT_ERROR)
                        + PayflowConstants.DELIMITER_NVP
                        + PayflowConstants.PARAM_RESPMSG
                        + PayflowConstants.SEPARATOR_NVP
                        + PayflowConstants.CommErrorMessages.get(PayflowConstants.E_INIT_ERROR)
                        + "host Address has not been initialised. Please make sure it is being set.";

                ErrorObject error = new ErrorObject(PayflowConstants.SEVERITY_FATAL, PayflowConstants.EMPTY_STRING, RespMessage);
                getContext().addError(error);
            }
        }

        if (proxyPort == 0) {
            proxyPort = SDKProperties.getProxyPort();
        }
        if (null == proxyAddress || proxyAddress.trim().length() == 0) {
            proxyAddress = SDKProperties.getProxyAddress();
            if (null == proxyAddress) {
                proxyAddress = PayflowConstants.EMPTY_STRING;
            }
        }
        if (null == proxyLogon || proxyLogon.trim().length() == 0) {
            proxyLogon = SDKProperties.getProxyLogin();
            if (null == proxyLogon) {
                proxyLogon = PayflowConstants.EMPTY_STRING;
            }
        }
        if (null == proxyPassword || proxyPassword.trim().length() == 0) {
            proxyPassword = SDKProperties.getProxyPassword();
            if (null == proxyPassword) {
                proxyPassword = PayflowConstants.EMPTY_STRING;
            }
        }
        Logger.getInstance().log("paypal.payflow.PayflowConnectionData.initSDKProperties(): Exiting", PayflowConstants.SEVERITY_DEBUG);
    }
}

